pkg/storegateway/partitioner.go

```diff
@@ -42,38 +43,41 @@ func newGapBasedPartitioner(maxGapBytes uint64, reg prometheus.Registerer) *gapB
 			Name: "cortex_bucket_store_partitioner_expanded_ranges_total",
 			Help: "Total number of byte ranges returned by the partitioner after they've been combined together to reduce the number of bucket API calls.",
 		}),
+		extendedRanges: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_bucket_store_partitioner_extended_ranges_total",
+			Help: "Total number of byte ranges that were not overlapping but were joined because they were closer than the configured maximum gap.",
+		}),
 	}
 }
 
 // Partition partitions length entries into n <= length ranges that cover all
 // input ranges by combining entries that are separated by reasonably small gaps.
 // It is used to combine multiple small ranges from object storage into bigger, more efficient/cheaper ones.
 func (g *gapBasedPartitioner) Partition(length int, rng func(int) (uint64, uint64)) []Part {
-	// Calculate the size of requested ranges.
-	requestedBytes := uint64(0)
-	for i := 0; i < length; i++ {
-		start, end := rng(i)
-		requestedBytes += end - start
-	}
-
 	// Run the upstream partitioner to compute the actual ranges that will be fetched.
-	parts := g.partition(length, rng)
+	parts, stats := g.partition(length, rng)
 
 	// Calculate the size of ranges that will be fetched.
 	expandedBytes := uint64(0)
 	for _, p := range parts {
 		expandedBytes += p.End - p.Start
 	}
 
-	g.requestedBytes.Add(float64(requestedBytes))
+	g.requestedBytes.Add(float64(stats.requestedRangesTotal))
 	g.expandedBytes.Add(float64(expandedBytes))
 	g.requestedRanges.Add(float64(length))
 	g.expandedRanges.Add(float64(len(parts)))
+	g.extendedRanges.Add(float64(stats.extendedNonOverlappingRanges))
 
 	return parts
 }
 
-func (g *gapBasedPartitioner) partition(length int, rng func(int) (uint64, uint64)) (parts []Part) {
+type partitionStats struct {
+	extendedNonOverlappingRanges int
+	requestedRangesTotal         uint64
```

Comment:
[nit] By "requested ranges" we mean the number of number of ranges, not their size.

```suggestion
	requestedBytesTotal         uint64
```